hazelcast/src/main/java/com/hazelcast/internal/partition/impl/InternalPartitionServiceImpl.java

```diff
@@ -1334,36 +1321,54 @@ public PartitionTableView createPartitionTableView() {
         }
     }
 
-    public boolean commitMigrationOnDestination(MigrationInfo migration, int newVersion, Address sender) {
+    @SuppressWarnings("checkstyle:npathcomplexity")
+    public boolean commitMigrationOnDestination(MigrationInfo migration, Address sender) {
         lock.lock();
         try {
             if (!validateSenderIsMaster(sender, "migration commit")) {
                 return false;
             }
 
-            Boolean versionCheckResult = checkIfPartitionStateVersionApplied(newVersion, sender);
-            if (versionCheckResult != null) {
-                return versionCheckResult;
-            }
+            int currentVersion = partitionStateManager.getVersion();
+            int initialVersion = migration.getInitialPartitionVersion();
+            int finalVersion = migration.getFinalPartitionVersion();
 
-            MigrationInfo activeMigrationInfo = migrationManager.getActiveMigration();
-            if (migration.equals(activeMigrationInfo)) {
-                InternalPartitionImpl partition = partitionStateManager.getPartitionImpl(migration.getPartitionId());
-                boolean added = migrationManager.addCompletedMigration(migration);
-                assert added : "Could not add completed migration on destination: " + migration;
-                migrationManager.applyMigration(partition, migration);
-                partitionStateManager.setVersion(newVersion);
-                activeMigrationInfo.setStatus(migration.getStatus());
-                migrationManager.finalizeMigration(migration);
-                if (logger.isFinestEnabled()) {
-                    logger.finest("Committed " + migration + " on destination with partition state version: " + newVersion);
+            if (finalVersion == currentVersion) {
+                if (logger.isFineEnabled()) {
+                    logger.fine("Already applied migration commit. Version: " + currentVersion + ", Master: " + sender);
                 }
                 return true;
             }
+            if (finalVersion < currentVersion) {
+                if (logger.isFineEnabled()) {
+                    logger.fine("Already applied migration commit. Local version: " + currentVersion
+                            + ", Master version: " + finalVersion + " Master: " + sender);
+                }
+                return false;
+            }
+            if (initialVersion != currentVersion) {
+                throw new IllegalStateException("Invalid migration commit! Expected version: " + initialVersion
```

Comment:
Actually, normally, I don't expect this branch to be executed. I just added as a safeguard. Because, destination accepts migration operation only when master's partition state version is equal to its local version (see `MigrationOperation`). When `commitMigrationOnDestination()` is called, either `initialVersion` should be equal to `currentVersion` or migration commit should be a stale operation (because of a retry).